ext/random/random.c

```diff
@@ -319,13 +319,13 @@ PHPAPI zend_object *php_random_engine_common_clone_object(zend_object *object)
 /* {{{ php_random_range */
 PHPAPI zend_long php_random_range(const php_random_algo *algo, php_random_status *status, zend_long min, zend_long max)
 {
-	zend_ulong umax = max - min;
+	zend_ulong umax = (zend_ulong) max - (zend_ulong) min;
```

Comment:
As it's public API, we might do a check here if the values are really positive or alternatively change argument types to zend_ulong

Reply:
Indeed.
In that case, the method of receiving only `zend_ulong umax` could be better, what do you think?

Reply:
I don't think any change is needed, see e.g. https://ideone.com/vTn10A

Reply:
@bukka 
> As it's public API, we might do a check here if the values are really positive or alternatively change argument types to zend_ulong
True, but I think it would make it harder to use as an API. I prefer the current internal casting scheme, but is there any advantage to being zend_ulong at argument time?

However, the PHP_RANDOM_ALGO_IS_DYNAMIC macro used in this function is apparently unnecessary. This has been removed in: 34dc686c2600ff5aa067f434d4046a39b7b010c2

Reply:
About similarity with `php_random_int`, the first thing it does (even before computing `umax`) is basically `if (min == max) return min;`, would it be "good" to do the same in `php_random_range`?

Reply:
> The values can also be _negative_, and I don't think any check/change is needed/better (thanks to C unsigned "modulo wrapping"), see e.g. https://ideone.com/Yak2Io (moreover it's consistent with existing `php_random_int`)

This is however implementation defined behavior (depends on compiler) which we should not rely on in general.

Reply:
See C99 and think the same in C11:

> 6.3.1.3 Signed and unsigned integers
1 When a value with integer type is converted to another integer type other than _Bool, if
the value can be represented by the new type, it is unchanged.
2 Otherwise, if the new type is unsigned, the value is converted by repeatedly adding or
subtracting one more than the maximum value that can be represented in the new type
until the value is in the range of the new type.49)
3 Otherwise, the new type is signed and the value cannot be represented in it; either the
result is implementation-defined or an implementation-defined signal is raised.

Reply:
That said I think we probably do such conversions in other places anyway. Just not sure if it's a good idea in general...

Reply:
@bukka See https://github.com/php/php-src/pull/8220#issuecomment-1073337678 where Nikita wrote:

> It's implementation-defined behavior, and all compiler implementations supported by PHP agree on the behavior.

Reply:
@bukka (and @iluuu1994): Anyway what is implementation-defined behavior is converting to _signed_ [from unsigned]. Converting to _**un**signed_ [from signed] is perfectly well-defined behavior (point 2 of above C99 6.3.1.3 quote, what I referred to as "modulo wrapping", which also applies when subtracting two unsigneds like here).

Reply:
Ah I see, my bad...